pkg/jit/jitdump.go

```diff
@@ -0,0 +1,386 @@
+type JRCodeDebugInfo struct {
+	Prefix   *JRPrefix     // the record header
+	CodeAddr uint64        // address of function for which the debug information is generated
+	NREntry  uint64        // number of debug entries for the function
+	Entries  []*DebugEntry // array of NREntry debug entries for the function
+}
+
+const jrCodeDebugInfoFixedSize int = jrPrefixSize + 16 // size of JRCodeDebugInfo fixed-sized fields
+
+// JRCodeUnwindingInfo represents a JITCodeUnwindingInfo record.
+type JRCodeUnwindingInfo struct {
+	Prefix         *JRPrefix // the record header
+	UnwindingSize  uint64    // the size in bytes of the unwinding data table at the end of the record
+	EHFrameHDRSize uint64    // the size in bytes of the DWARF EH Frame Header at the start of the unwinding data table at the end of the record
+	MappedSize     uint64    // the size of the unwinding data mapped in memory
+	UnwindingData  []byte    // an array of unwinding data, consisting of the EH Frame Header, followed by the actual EH Frame
+}
+
+const jrCodeUnwindingInfoFixedSize int = jrPrefixSize + 24 // size of JRCodeUnwindingInfo fixed-sized fields
+
+// JITDump represents the loaded jitdump.
+type JITDump struct {
+	Header        *JITHeader             // the jitdump file header
+	CodeLoads     []*JRCodeLoad          // JITCodeLoad records
+	CodeMoves     []*JRCodeMove          // JITCodeMove records
+	DebugInfo     []*JRCodeDebugInfo     // JITCodeDebugInfo records
+	UnwindingInfo []*JRCodeUnwindingInfo // JITCodeUnwindingInfo records
+}
+
+// jitDumpParser is used to parse a jitdump file.
+type jitDumpParser struct {
+	logger     log.Logger
+	buf        *bufio.Reader
+	endianness binary.ByteOrder
+}
+
+// newParser initializes a jitDumpParser.
+func newParser(logger log.Logger, rd io.Reader) (*jitDumpParser, error) {
+	p := &jitDumpParser{
+		logger: logger,
+	}
+
+	p.buf = bufio.NewReader(rd)
+
+	magic, err := p.buf.Peek(4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read magic number from jitdump file: %w", err)
+	}
+
+	switch {
+	case bytes.Equal(magic, []byte{'J', 'i', 'T', 'D'}):
+		p.endianness = binary.BigEndian
+	case bytes.Equal(magic, []byte{'D', 'T', 'i', 'J'}):
+		p.endianness = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("failed to detect JIT dump endianness from %#x magic number: %w", magic, err)
+	}
+
+	return p, nil
+}
+
+// read reads structured binary data from the jitdump file into data.
+func (p *jitDumpParser) read(data any) error {
+	return binary.Read(p.buf, p.endianness, data)
+}
+
+func (p *jitDumpParser) parseJITHeader() (*JITHeader, error) {
+	header := &JITHeader{}
+	if err := p.read(header); err != nil {
+		return nil, fmt.Errorf("failed to read jitdump header: %w", err)
+	}
+
+	if header.Version > JITHeaderVersion {
+		return nil, fmt.Errorf("wrong jitdump version: %d (expected: %d)", header.Version, JITHeaderVersion)
+	}
+
+	return header, nil
+}
+
+func (p *jitDumpParser) parseJRCodeLoad(prefix *JRPrefix) *JRCodeLoad {
```

Comment:
These helper functions should return errors with context, and the caller can log.

Reply:
We can use something like https://pkg.go.dev/go.uber.org/multierr

Reply:
I wonder if it is possible to do with `fmt.Errorf()`, I've never tried, so I will try

Reply:
Wrapping is possible but combining multiple errors is not. We use similar packages elsewhere.